File: 1302_DeepestLeavesSum/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
// BFS Approach
class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
        if(!root) return 0;
        
        queue<TreeNode*> q;
        q.push(root);
        int res = 0;
        while(!q.empty()){
            int size = q.size();
            res = 0;
            for(int i = 0; i < size; i++){
                TreeNode* root = q.front(); q.pop();
                res += root->val;
                if(root->left != nullptr) q.push(root->left);
                if(root->right != nullptr) q.push(root->right);
            }
        }
        return res;
    }
};

// DFS Approach

class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
        if(!dfs(root, 0)) return 0;
        return sumMap[maxDepth];
    }
    
private:
    int maxDepth{0};
    unordered_map<int, int> sumMap;
    
    bool dfs(TreeNode* node, int curLevel){
        if(!node) return false; 
        
        if( curLevel >= maxDepth ){
            maxDepth = curLevel;
            sumMap[maxDepth] += node->val;
        }
        dfs(node->left, curLevel + 1);
        dfs(node->right, curLevel + 1);
        
        return true;
    }
};   

// or can pass down maxDepth, sum as references

File: 1302_DeepestLeavesSum/main.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

//BFS Approach
class Solution {
    public int deepestLeavesSum(TreeNode root) {
        int res = 0;
        if(root == null) return res;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        while(!q.isEmpty()){
            int curSize = q.size();
            res = 0;
            for(int i = 0; i < curSize; i++){
                TreeNode curNode = q.poll();
                res += curNode.val;
                if(curNode.left != null) q.offer(curNode.left);
                if(curNode.right != null) q.offer(curNode.right);
            }
        }
        return res;
    }
}


// DFS Approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    private int maxDepth = 0;
    private int sum = 0;
    
    public int deepestLeavesSum(TreeNode root) {
        if(root == null) return 0;
        dfs(root, 1);
        return sum;
    }
    
    private void dfs(TreeNode root, int curDepth){
        if(root == null) return;
        
        if(curDepth == maxDepth) sum += root.val;
        if(curDepth > maxDepth){
            sum = root.val;
            maxDepth = curDepth;
        }
        
        dfs(root.left, curDepth + 1);
        dfs(root.right, curDepth + 1);
    }
}
